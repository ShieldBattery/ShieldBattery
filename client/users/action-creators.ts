import { getErrorStack } from '../../common/errors'
import { apiUrl, urlPath } from '../../common/urls'
import { SbPermissions } from '../../common/users/permissions'
import { GetRelationshipsResponse } from '../../common/users/relationships'
import {
  AdminBanUserRequest,
  AdminBanUserResponse,
  AdminGetBansResponse,
  AdminGetPermissionsResponse,
  AdminGetUserIpsResponse,
  AdminUpdatePermissionsRequest,
  GetBatchUserInfoResponse,
  GetUserProfileResponse,
  GetUserRankingHistoryResponse,
  SbUserId,
  SearchMatchHistoryResponse,
} from '../../common/users/sb-user'
import { ThunkAction } from '../dispatch-registry'
import logger from '../logging/logger'
import { push, replace } from '../navigation/routing'
import { RequestHandlingSpec, abortableThunk } from '../network/abortable-thunk'
import { MicrotaskBatchRequester } from '../network/batch-requests'
import { encodeBodyAsParams, fetchJson } from '../network/fetch'
import { RequestCoalescer } from '../network/request-coalescer'
import { UserProfileSubPage } from './user-profile-sub-page'

/**
 * Navigates to a specific user's profile (and optionally, a specific tab within that).
 */
export function navigateToUserProfile(
  userId: SbUserId,
  username: string,
  tab?: UserProfileSubPage,
  transitionFn = push,
) {
  transitionFn(urlPath`/users/${userId}/${username}/${tab ?? ''}`)
}

/**
 * Corrects the URL for a specific user's profile if it is already being viewed. This is meant to be
 * used when the client arrived on the page but the username doesn't match what we have stored for
 * their user ID.
 */
export function correctUsernameForProfile(
  userId: SbUserId,
  username: string,
  tab?: UserProfileSubPage,
) {
  replace(urlPath`/users/${userId}/${username}/${tab ?? ''}`)
}

const viewUserProfileRequestCoalescer = new RequestCoalescer<SbUserId>()

/**
 * Signals that a specific user's profile is being viewed. If we don't have a local copy of that
 * user's profile data already, it will be retrieved from the server.
 */
export function viewUserProfile(userId: SbUserId, spec: RequestHandlingSpec<void>): ThunkAction {
  return abortableThunk(spec, async dispatch => {
    await viewUserProfileRequestCoalescer.makeRequest(
      userId,
      spec.signal,
      async (signal: AbortSignal) => {
        dispatch({
          type: '@users/getUserProfile',
          payload: await fetchJson<GetUserProfileResponse>(apiUrl`users/${userId}/profile`, {
            signal,
          }),
        })
      },
    )
  })
}

const MAX_BATCH_INFO_REQUESTS = 20

const infoBatchRequester = new MicrotaskBatchRequester<SbUserId>(
  MAX_BATCH_INFO_REQUESTS,
  (dispatch, items) => {
    const params = items.map(u => urlPath`u=${u}`).join('&')
    const promise = fetchJson<GetBatchUserInfoResponse>(apiUrl`users/batch-info` + '?' + params)
    dispatch({
      type: '@users/getBatchUserInfo',
      payload: promise,
      meta: {
        userIds: items,
      },
    })

    return promise
  },
  err => {
    logger.error('error while batch requesting user info: ' + getErrorStack(err))
  },
)

export function getBatchUserInfo(userId: SbUserId): ThunkAction {
  return (dispatch, getState) => {
    const {
      users: { byId },
    } = getState()

    if (!byId.has(userId)) {
      infoBatchRequester.request(dispatch, userId)
    }
  }
}

export function searchMatchHistory(
  userId: SbUserId,
  offset: number,
  spec: RequestHandlingSpec<SearchMatchHistoryResponse>,
): ThunkAction {
  return abortableThunk(spec, async dispatch => {
    const result = await fetchJson<SearchMatchHistoryResponse>(
      apiUrl`users/${userId}/match-history?offset=${offset}`,
      {
        signal: spec.signal,
      },
    )

    dispatch({
      type: '@users/searchMatchHistory',
      payload: result,
      meta: { userId },
    })

    return result
  })
}

export function adminGetUserPermissions(
  userId: SbUserId,
  spec: RequestHandlingSpec<AdminGetPermissionsResponse>,
): ThunkAction {
  return abortableThunk(spec, async () => {
    return await fetchJson<AdminGetPermissionsResponse>(apiUrl`admin/users/${userId}/permissions`, {
      signal: spec.signal,
    })
  })
}

export function adminUpdateUserPermissions(
  userId: SbUserId,
  permissions: SbPermissions,
  spec: RequestHandlingSpec<void>,
): ThunkAction {
  return abortableThunk(spec, async () => {
    const body: AdminUpdatePermissionsRequest = {
      permissions,
    }
    return await fetchJson<void>(apiUrl`admin/users/${userId}/permissions`, {
      method: 'POST',
      body: JSON.stringify(body),
      signal: spec.signal,
    })
  })
}

export function adminGetUserBanHistory(
  userId: SbUserId,
  spec: RequestHandlingSpec<AdminGetBansResponse>,
): ThunkAction {
  return abortableThunk(spec, async dispatch => {
    const res = await fetchJson<AdminGetBansResponse>(apiUrl`admin/users/${userId}/bans`, {
      signal: spec.signal,
    })
    dispatch({ type: '@users/adminGetUserBanHistory', payload: res })

    return res
  })
}

export function adminBanUser(
  { userId, banLengthHours, reason }: { userId: SbUserId; banLengthHours: number; reason?: string },
  spec: RequestHandlingSpec<AdminBanUserResponse>,
): ThunkAction {
  return abortableThunk(spec, async dispatch => {
    const res = await fetchJson<AdminBanUserResponse>(apiUrl`admin/users/${userId}/bans`, {
      method: 'POST',
      body: encodeBodyAsParams<AdminBanUserRequest>({
        banLengthHours,
        reason,
      }),
      signal: spec.signal,
    })
    dispatch({ type: '@users/adminBanUser', payload: res })

    return res
  })
}

export function adminGetUserIps(
  userId: SbUserId,
  spec: RequestHandlingSpec<AdminGetUserIpsResponse>,
): ThunkAction {
  return abortableThunk(spec, async dispatch => {
    const res = await fetchJson<AdminGetUserIpsResponse>(apiUrl`admin/users/${userId}/ips`, {
      signal: spec.signal,
    })
    dispatch({ type: '@users/adminGetUserIps', payload: res })

    return res
  })
}

export function getRelationshipsIfNeeded(spec: RequestHandlingSpec): ThunkAction {
  return abortableThunk(spec, async (dispatch, getState) => {
    const {
      auth: { self },
      relationships,
    } = getState()
    if (relationships.loaded || !self) {
      return
    }

    const result = await fetchJson<GetRelationshipsResponse>(
      apiUrl`users/${self.user.id}/relationships`,
      { signal: spec.signal },
    )
    dispatch({ type: '@users/getRelationships', payload: result })
  })
}

export function sendFriendRequest(toId: SbUserId, spec: RequestHandlingSpec): ThunkAction {
  return abortableThunk(spec, async () => {
    await fetchJson<void>(apiUrl`users/${toId}/relationships/friend-requests`, {
      method: 'POST',
      signal: spec.signal,
    })
  })
}

export function removeFriendRequest(toId: SbUserId, spec: RequestHandlingSpec): ThunkAction {
  return abortableThunk(spec, async (_, getState) => {
    const {
      auth: { self },
    } = getState()

    await fetchJson<void>(apiUrl`users/${toId}/relationships/friend-requests/${self!.user.id}`, {
      method: 'DELETE',
      signal: spec.signal,
    })
  })
}

export function acceptFriendRequest(fromId: SbUserId, spec: RequestHandlingSpec): ThunkAction {
  return abortableThunk(spec, async (_, getState) => {
    const {
      auth: { self },
    } = getState()

    await fetchJson<void>(apiUrl`users/${self!.user.id}/relationships/friends/${fromId}`, {
      method: 'POST',
      signal: spec.signal,
    })
  })
}

export function declineFriendRequest(fromId: SbUserId, spec: RequestHandlingSpec): ThunkAction {
  return abortableThunk(spec, async (_, getState) => {
    const {
      auth: { self },
    } = getState()

    await fetchJson<void>(apiUrl`users/${self!.user.id}/relationships/friend-requests/${fromId}`, {
      method: 'DELETE',
      signal: spec.signal,
    })
  })
}

export function removeFriend(targetId: SbUserId, spec: RequestHandlingSpec): ThunkAction {
  return abortableThunk(spec, async (_, getState) => {
    const {
      auth: { self },
    } = getState()

    await fetchJson<void>(apiUrl`users/${self!.user.id}/relationships/friends/${targetId}`, {
      method: 'DELETE',
      signal: spec.signal,
    })
  })
}

export function blockUser(targetId: SbUserId, spec: RequestHandlingSpec): ThunkAction {
  return abortableThunk(spec, async (_, getState) => {
    const {
      auth: { self },
    } = getState()

    await fetchJson<void>(apiUrl`users/${self!.user.id}/relationships/blocks/${targetId}`, {
      method: 'POST',
      signal: spec.signal,
    })
  })
}

export function unblockUser(targetId: SbUserId, spec: RequestHandlingSpec): ThunkAction {
  return abortableThunk(spec, async (_, getState) => {
    const {
      auth: { self },
    } = getState()

    await fetchJson<void>(apiUrl`users/${self!.user.id}/relationships/blocks/${targetId}`, {
      method: 'DELETE',
      signal: spec.signal,
    })
  })
}

export function getUserRankingHistory(
  userId: SbUserId,
  spec: RequestHandlingSpec<GetUserRankingHistoryResponse>,
): ThunkAction {
  return abortableThunk(spec, async dispatch => {
    const result = await fetchJson<GetUserRankingHistoryResponse>(
      apiUrl`users/${userId}/ranking-history`,
      {
        signal: spec.signal,
      },
    )

    dispatch({
      type: '@users/getRankingHistory',
      payload: result,
      meta: { userId },
    })

    return result
  })
}
