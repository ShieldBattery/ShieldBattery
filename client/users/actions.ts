import {
  FriendActivityStatusUpdateEvent,
  GetRelationshipsResponse,
  UserRelationshipJson,
} from '../../common/users/relationships'
import {
  AdminBanUserResponse,
  AdminGetBansResponse,
  AdminGetUserIpsResponse,
  GetBatchUserInfoResponse,
  GetUserProfileResponse,
  SbUser,
  SbUserId,
} from '../../common/users/sb-user'

export type UserActions =
  | GetUserProfile
  | GetBatchUserInfo
  | LoadUsers
  | AdminGetUserBanHistory
  | AdminBanUser
  | AdminGetUserIps
  | GetRelationships
  | UpsertUserRelationship
  | DeleteUserRelationship
  | UpdateFriendActivityStatus

export interface GetUserProfile {
  type: '@users/getUserProfile'
  payload: GetUserProfileResponse
}

export type GetBatchUserInfo =
  | {
      type: '@users/getBatchUserInfo'
      payload: GetBatchUserInfoResponse
      error?: false
      meta: {
        userIds: ReadonlyArray<SbUserId>
      }
    }
  | {
      type: '@users/getBatchUserInfo'
      payload: Error
      error: true
      meta: {
        userIds: ReadonlyArray<SbUserId>
      }
    }

export interface LoadUsers {
  type: '@users/loadUsers'
  payload: SbUser[]
}

export interface AdminGetUserBanHistory {
  type: '@users/adminGetUserBanHistory'
  payload: AdminGetBansResponse
}

export interface AdminBanUser {
  type: '@users/adminBanUser'
  payload: AdminBanUserResponse
}

export interface AdminGetUserIps {
  type: '@users/adminGetUserIps'
  payload: AdminGetUserIpsResponse
}

export interface GetRelationships {
  type: '@users/getRelationships'
  payload: GetRelationshipsResponse
}

export interface UpsertUserRelationship {
  type: '@users/upsertRelationship'
  payload: {
    relationship: UserRelationshipJson
  }
  meta: { selfId: SbUserId }
}

export interface DeleteUserRelationship {
  type: '@users/deleteRelationship'
  payload: {
    targetUser: SbUserId
  }
}

export interface UpdateFriendActivityStatus {
  type: '@users/updateFriendActivityStatus'
  payload: FriendActivityStatusUpdateEvent
}
