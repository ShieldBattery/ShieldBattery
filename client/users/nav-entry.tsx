import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import { Avatar } from '../avatars/avatar'
import { useButtonState } from '../material/button'
import { buttonReset } from '../material/button-reset'
import { Ripple } from '../material/ripple'
import { AnimatedExpandIcon } from '../styles/animated-expand-icon'
import { colorTextFaint, colorTextPrimary } from '../styles/colors'
import { cabin, singleLine } from '../styles/typography'

const Container = styled.button`
  ${buttonReset};
  width: 100%;
  height: 72px;
  margin-top: 16px;
  padding: 0px 12px 0px 16px;

  display: flex;
  flex-direction: row;
  align-items: center;

  border-radius: 0 2px 0 0;
  cursor: pointer;
  text-align: left;

  & ${AnimatedExpandIcon} {
    color: ${colorTextFaint};
  }

  &:hover,
  &:active {
    & ${AnimatedExpandIcon} {
      color: ${colorTextPrimary};
    }
  }
`

const StyledAvatar = styled(Avatar)`
  width: 40px;
  height: 40px;
  display: block;
  margin: auto;
  overflow: hidden;
`

const User = styled.div`
  ${cabin}
  ${singleLine};

  width: 140px;
  margin-left: 16px;
  flex-grow: 1;

  font-size: 24px;
  font-weight: 500;
  line-height: 36px;
  letter-spacing: 0.5px;
`

export interface ProfileNavEntryProps {
  user: string
  onProfileEntryClick: (event: React.MouseEvent) => void
  profileMenuOpen: boolean
}

const ProfileNavEntry = React.forwardRef(
  (
    { user, onProfileEntryClick, profileMenuOpen }: ProfileNavEntryProps,
    ref: React.ForwardedRef<HTMLButtonElement>,
  ) => {
    const [buttonProps, rippleRef] = useButtonState({ onClick: onProfileEntryClick })

    return (
      <Container ref={ref} {...buttonProps}>
        <StyledAvatar user={user} />
        <User>{user}</User>
        <AnimatedExpandIcon $pointUp={!profileMenuOpen} />
        <Ripple ref={rippleRef} />
      </Container>
    )
  },
)

ProfileNavEntry.propTypes = {
  onProfileEntryClick: PropTypes.func.isRequired,
  user: PropTypes.string.isRequired,
}

export default ProfileNavEntry
