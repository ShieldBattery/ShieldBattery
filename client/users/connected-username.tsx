import React, { useEffect } from 'react'
import styled from 'styled-components'
import { SbUserId } from '../../common/users/sb-user'
import { useAppDispatch, useAppSelector } from '../redux-hooks'
import { colorDividers } from '../styles/colors'
import { getBatchUserInfo } from './action-creators'
import { ConnectedUserContextMenu, MenuItemCategory } from './user-context-menu'
import { useUserOverlays } from './user-overlays'
import { ConnectedUserProfileOverlay } from './user-profile-overlay'

const Username = styled.span`
  &:hover {
    cursor: pointer;
    text-decoration: underline;
  }
`

const LoadingName = styled.span`
  margin-right: 0.25em;
  background-color: ${colorDividers};
  border-radius: 2px;
`

export interface ConnectedUsernameProps {
  className?: string
  /** The user to show the corresponding name for. */
  userId: SbUserId
  /** A string to show before the username, e.g. '@' for mentions. */
  prefix?: string
  /**
   * An optional callback that will be called before the normal `onClick` handling. If the click
   * was handled by the callback, it should return `true` to indicate the normal behavior should
   * not occur.
   */
  filterClick?: (userId: SbUserId, e: React.MouseEvent) => boolean
  /**
   * An optional function that will be called when rendering menu items. If provided, the value
   * returned from this function will be used as the `children` of the menu. Mutating the input
   * value and returning it is okay.
   */
  modifyMenuItems?: (
    userId: SbUserId,
    items: Map<MenuItemCategory, React.ReactNode[]>,
    onMenuClose: (event?: MouseEvent) => void,
  ) => Map<MenuItemCategory, React.ReactNode[]>
}

/**
 * A component which displays a clickable username, displaying the user's profile overlay or
 * context menu when clicked.
 */
export function ConnectedUsername({
  className,
  userId,
  prefix = '',
  filterClick,
  modifyMenuItems,
}: ConnectedUsernameProps) {
  const dispatch = useAppDispatch()
  const user = useAppSelector(s => s.users.byId.get(userId))

  useEffect(() => {
    dispatch(getBatchUserInfo(userId))
  }, [dispatch, userId])

  const { clickableElemRef, profileOverlayProps, contextMenuProps, onClick, onContextMenu } =
    useUserOverlays<HTMLSpanElement>({
      userId,
      profileAnchorX: 'right',
      profileAnchorY: 'top',
      profileOriginX: 'left',
      profileOriginY: 'top',
      profileOffsetX: 4,
      filterClick,
      modifyMenuItems,
    })

  const username = user?.name ?? (
    <LoadingName aria-label={'Username loadingâ€¦'}>
      &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;
    </LoadingName>
  )

  return (
    <>
      <ConnectedUserProfileOverlay {...profileOverlayProps} />
      <ConnectedUserContextMenu {...contextMenuProps} />

      <Username
        ref={clickableElemRef}
        className={className}
        onClick={onClick}
        onContextMenu={onContextMenu}>
        {prefix}
        {username}
      </Username>
    </>
  )
}
