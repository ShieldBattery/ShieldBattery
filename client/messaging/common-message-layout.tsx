import React, { useMemo } from 'react'
import styled from 'styled-components'
import { amberA100 } from '../styles/colors'
import { body2 } from '../styles/typography'
import { ConnectedUsername } from './connected-username'
import {
  InfoImportant,
  SeparatedInfoMessage,
  Separator,
  TimestampMessageLayout,
} from './message-layout'

const URL_REGEX = /(?<g1>https?:\/\/)(?:[^\s)"\].]|(?:\.(?=\S))|(?<=\k<g1>.*\([^)]*)\)){2,}/gi

const newDayFormat = new Intl.DateTimeFormat(navigator.language, {
  year: 'numeric',
  month: 'long',
  day: '2-digit',
})

const Username = styled.span`
  ${body2};

  margin-right: 8px;

  color: ${amberA100};
  line-height: inherit;
`

const Text = styled.span`
  line-height: inherit;
  word-wrap: break-word;
  overflow-wrap: break-word;
  overflow: hidden;
`

export function ParsedText({ text }: { text: string }) {
  const parsedText = useMemo(() => {
    const matches = text.matchAll(URL_REGEX)
    const elements = []
    let lastIndex = 0

    for (const match of matches) {
      // Insert preceding text, if any
      if (match.index! > lastIndex) {
        elements.push(text.substring(lastIndex, match.index))
      }

      elements.push(
        // TODO(2Pac): Show a warning message about opening untrusted links
        <a key={match.index} href={match[0]} target='_blank' rel='noopener'>
          {match[0]}
        </a>,
      )

      lastIndex = match.index! + match[0].length
    }

    // Insert remaining text, if any
    if (text.length > lastIndex) {
      elements.push(text.substring(lastIndex))
    }

    return elements
  }, [text])

  return <Text>{parsedText}</Text>
}

export const TextMessageDisplay = React.memo<{ userId: number; time: number; text: string }>(
  props => {
    const { userId, time, text } = props
    return (
      <TimestampMessageLayout time={time}>
        <Username>
          <ConnectedUsername userId={userId} />
        </Username>
        <Separator aria-hidden={true}>{': '}</Separator>
        <ParsedText text={text} />
      </TimestampMessageLayout>
    )
  },
)

export const NewDayMessage = React.memo<{ time: number }>(props => {
  const { time } = props
  return (
    <SeparatedInfoMessage>
      <span>
        Day changed to <InfoImportant>{newDayFormat.format(time)}</InfoImportant>
      </span>
    </SeparatedInfoMessage>
  )
})
