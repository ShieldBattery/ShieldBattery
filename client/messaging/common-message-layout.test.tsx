import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { makeSbUserId } from '../../common/users/sb-user'
import { TextMessage } from './common-message-layout'

const selfUserId = makeSbUserId(1)
const userId = makeSbUserId(2)

describe('client/messaging/common-message-layout/TextMessage', () => {
  const renderer = ShallowRenderer.createRenderer()
  const doRender = (text: string): React.ReactElement => {
    renderer.render(<TextMessage userId={userId} selfUserId={selfUserId} time={0} text={text} />)
    return renderer.getRenderOutput()
  }

  test('message as a normal text', () => {
    expect(doRender('This is test message')).toMatchInlineSnapshot(`
      <TimestampMessageLayout
        highlighted={false}
        time={0}
      >
        <common-message-layout__Username>
          <ConnectedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            userId={2}
          />
        </common-message-layout__Username>
        <message-layout__Separator>
          : 
        </message-layout__Separator>
        <common-message-layout__Text>
          This is test message
        </common-message-layout__Text>
      </TimestampMessageLayout>
    `)
  })

  test('message with a link', () => {
    expect(doRender('here is a link http://www.example.com')).toMatchInlineSnapshot(`
      <TimestampMessageLayout
        highlighted={false}
        time={0}
      >
        <common-message-layout__Username>
          <ConnectedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            userId={2}
          />
        </common-message-layout__Username>
        <message-layout__Separator>
          : 
        </message-layout__Separator>
        <common-message-layout__Text>
          here is a link 
          <ExternalLink
            href="http://www.example.com"
          >
            http://www.example.com
          </ExternalLink>
        </common-message-layout__Text>
      </TimestampMessageLayout>
    `)
  })

  test('message with a mention', () => {
    expect(doRender('hey <@123>')).toMatchInlineSnapshot(`
      <TimestampMessageLayout
        highlighted={false}
        time={0}
      >
        <common-message-layout__Username>
          <ConnectedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            userId={2}
          />
        </common-message-layout__Username>
        <message-layout__Separator>
          : 
        </message-layout__Separator>
        <common-message-layout__Text>
          hey
           
          <common-message-layout__MentionedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            prefix="@"
            userId={123}
          />
        </common-message-layout__Text>
      </TimestampMessageLayout>
    `)
  })

  test('message with a link before a mention', () => {
    expect(doRender('http://www.example.com go here <@123>')).toMatchInlineSnapshot(`
      <TimestampMessageLayout
        highlighted={false}
        time={0}
      >
        <common-message-layout__Username>
          <ConnectedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            userId={2}
          />
        </common-message-layout__Username>
        <message-layout__Separator>
          : 
        </message-layout__Separator>
        <common-message-layout__Text>
          <ExternalLink
            href="http://www.example.com"
          >
            http://www.example.com
          </ExternalLink>
           go here
           
          <common-message-layout__MentionedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            prefix="@"
            userId={123}
          />
        </common-message-layout__Text>
      </TimestampMessageLayout>
    `)
  })

  test('message with a link between mentions', () => {
    expect(doRender('hey <@123> see http://www.example.com go here <@123>')).toMatchInlineSnapshot(`
      <TimestampMessageLayout
        highlighted={false}
        time={0}
      >
        <common-message-layout__Username>
          <ConnectedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            userId={2}
          />
        </common-message-layout__Username>
        <message-layout__Separator>
          : 
        </message-layout__Separator>
        <common-message-layout__Text>
          hey
           
          <common-message-layout__MentionedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            prefix="@"
            userId={123}
          />
           see 
          <ExternalLink
            href="http://www.example.com"
          >
            http://www.example.com
          </ExternalLink>
           go here
           
          <common-message-layout__MentionedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            prefix="@"
            userId={123}
          />
        </common-message-layout__Text>
      </TimestampMessageLayout>
    `)
  })

  test('message with a mention before a link', () => {
    expect(doRender('<@123> go to http://www.example.com')).toMatchInlineSnapshot(`
      <TimestampMessageLayout
        highlighted={false}
        time={0}
      >
        <common-message-layout__Username>
          <ConnectedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            userId={2}
          />
        </common-message-layout__Username>
        <message-layout__Separator>
          : 
        </message-layout__Separator>
        <common-message-layout__Text>
          
          <common-message-layout__MentionedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            prefix="@"
            userId={123}
          />
           go to 
          <ExternalLink
            href="http://www.example.com"
          >
            http://www.example.com
          </ExternalLink>
        </common-message-layout__Text>
      </TimestampMessageLayout>
    `)
  })

  test('message with a mention between links', () => {
    expect(doRender('http://www.example.com go here <@123> or here http://www.example.com'))
      .toMatchInlineSnapshot(`
      <TimestampMessageLayout
        highlighted={false}
        time={0}
      >
        <common-message-layout__Username>
          <ConnectedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            userId={2}
          />
        </common-message-layout__Username>
        <message-layout__Separator>
          : 
        </message-layout__Separator>
        <common-message-layout__Text>
          <ExternalLink
            href="http://www.example.com"
          >
            http://www.example.com
          </ExternalLink>
           go here
           
          <common-message-layout__MentionedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            prefix="@"
            userId={123}
          />
           or here 
          <ExternalLink
            href="http://www.example.com"
          >
            http://www.example.com
          </ExternalLink>
        </common-message-layout__Text>
      </TimestampMessageLayout>
    `)
  })

  test('message with a mention of self user', () => {
    expect(doRender('Hey <@1>')).toMatchInlineSnapshot(`
      <TimestampMessageLayout
        highlighted={true}
        time={0}
      >
        <common-message-layout__Username>
          <ConnectedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            userId={2}
          />
        </common-message-layout__Username>
        <message-layout__Separator>
          : 
        </message-layout__Separator>
        <common-message-layout__Text>
          Hey
           
          <common-message-layout__MentionedUsername
            filterClick={[Function]}
            modifyMenuItems={[Function]}
            prefix="@"
            userId={1}
          />
        </common-message-layout__Text>
      </TimestampMessageLayout>
    `)
  })
})
