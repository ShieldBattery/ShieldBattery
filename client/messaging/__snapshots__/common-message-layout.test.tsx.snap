// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client/messaging/common-message-layout/TextMessage message as a normal text 1`] = `
<React.Fragment>
  <TimestampMessageLayout
    active={false}
    highlighted={false}
    onContextMenu={[Function]}
    time={0}
  >
    <common-message-layout__Username>
      <ConnectedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        userId={2}
      />
    </common-message-layout__Username>
    <message-layout__Separator>
      : 
    </message-layout__Separator>
    <common-message-layout__Text>
      This is test message
    </common-message-layout__Text>
  </TimestampMessageLayout>
</React.Fragment>
`;

exports[`client/messaging/common-message-layout/TextMessage message with a link 1`] = `
<React.Fragment>
  <TimestampMessageLayout
    active={false}
    highlighted={false}
    onContextMenu={[Function]}
    time={0}
  >
    <common-message-layout__Username>
      <ConnectedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        userId={2}
      />
    </common-message-layout__Username>
    <message-layout__Separator>
      : 
    </message-layout__Separator>
    <common-message-layout__Text>
      here is a link 
      <ExternalLink
        href="http://www.example.com"
      >
        http://www.example.com
      </ExternalLink>
    </common-message-layout__Text>
  </TimestampMessageLayout>
</React.Fragment>
`;

exports[`client/messaging/common-message-layout/TextMessage message with a link before a mention 1`] = `
<React.Fragment>
  <TimestampMessageLayout
    active={false}
    highlighted={false}
    onContextMenu={[Function]}
    time={0}
  >
    <common-message-layout__Username>
      <ConnectedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        userId={2}
      />
    </common-message-layout__Username>
    <message-layout__Separator>
      : 
    </message-layout__Separator>
    <common-message-layout__Text>
      <ExternalLink
        href="http://www.example.com"
      >
        http://www.example.com
      </ExternalLink>
       go here
       
      <common-message-layout__MentionedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        prefix="@"
        userId={123}
      />
    </common-message-layout__Text>
  </TimestampMessageLayout>
</React.Fragment>
`;

exports[`client/messaging/common-message-layout/TextMessage message with a link between mentions 1`] = `
<React.Fragment>
  <TimestampMessageLayout
    active={false}
    highlighted={false}
    onContextMenu={[Function]}
    time={0}
  >
    <common-message-layout__Username>
      <ConnectedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        userId={2}
      />
    </common-message-layout__Username>
    <message-layout__Separator>
      : 
    </message-layout__Separator>
    <common-message-layout__Text>
      hey
       
      <common-message-layout__MentionedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        prefix="@"
        userId={123}
      />
       see 
      <ExternalLink
        href="http://www.example.com"
      >
        http://www.example.com
      </ExternalLink>
       go here
       
      <common-message-layout__MentionedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        prefix="@"
        userId={123}
      />
    </common-message-layout__Text>
  </TimestampMessageLayout>
</React.Fragment>
`;

exports[`client/messaging/common-message-layout/TextMessage message with a mention 1`] = `
<React.Fragment>
  <TimestampMessageLayout
    active={false}
    highlighted={false}
    onContextMenu={[Function]}
    time={0}
  >
    <common-message-layout__Username>
      <ConnectedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        userId={2}
      />
    </common-message-layout__Username>
    <message-layout__Separator>
      : 
    </message-layout__Separator>
    <common-message-layout__Text>
      hey
       
      <common-message-layout__MentionedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        prefix="@"
        userId={123}
      />
    </common-message-layout__Text>
  </TimestampMessageLayout>
</React.Fragment>
`;

exports[`client/messaging/common-message-layout/TextMessage message with a mention before a link 1`] = `
<React.Fragment>
  <TimestampMessageLayout
    active={false}
    highlighted={false}
    onContextMenu={[Function]}
    time={0}
  >
    <common-message-layout__Username>
      <ConnectedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        userId={2}
      />
    </common-message-layout__Username>
    <message-layout__Separator>
      : 
    </message-layout__Separator>
    <common-message-layout__Text>
      
      <common-message-layout__MentionedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        prefix="@"
        userId={123}
      />
       go to 
      <ExternalLink
        href="http://www.example.com"
      >
        http://www.example.com
      </ExternalLink>
    </common-message-layout__Text>
  </TimestampMessageLayout>
</React.Fragment>
`;

exports[`client/messaging/common-message-layout/TextMessage message with a mention between links 1`] = `
<React.Fragment>
  <TimestampMessageLayout
    active={false}
    highlighted={false}
    onContextMenu={[Function]}
    time={0}
  >
    <common-message-layout__Username>
      <ConnectedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        userId={2}
      />
    </common-message-layout__Username>
    <message-layout__Separator>
      : 
    </message-layout__Separator>
    <common-message-layout__Text>
      <ExternalLink
        href="http://www.example.com"
      >
        http://www.example.com
      </ExternalLink>
       go here
       
      <common-message-layout__MentionedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        prefix="@"
        userId={123}
      />
       or here 
      <ExternalLink
        href="http://www.example.com"
      >
        http://www.example.com
      </ExternalLink>
    </common-message-layout__Text>
  </TimestampMessageLayout>
</React.Fragment>
`;

exports[`client/messaging/common-message-layout/TextMessage message with a mention of self user 1`] = `
<React.Fragment>
  <TimestampMessageLayout
    active={false}
    highlighted={true}
    onContextMenu={[Function]}
    time={0}
  >
    <common-message-layout__Username>
      <ConnectedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        userId={2}
      />
    </common-message-layout__Username>
    <message-layout__Separator>
      : 
    </message-layout__Separator>
    <common-message-layout__Text>
      Hey
       
      <common-message-layout__MentionedUsername
        filterClick={[Function]}
        modifyMenuItems={[Function]}
        prefix="@"
        userId={1}
      />
    </common-message-layout__Text>
  </TimestampMessageLayout>
</React.Fragment>
`;
